File: src/cpp/pipeline.cpp
#include <cmath>
#include <deque>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>

// Rolling window size for moving average calculation
constexpr int ROLLING_WINDOW = 3;

// Threshold (%) to flag price anomalies (sudden jumps)
constexpr double ANOMALY_THRESHOLD_PCT = 0.8;

// Holds aggregated state for each symbol during processing
struct SymbolState {
    long long total_volume = 0;        // Total traded volume
    double total_notional = 0.0;       // Sum of (price * volume) for VWAP
    std::deque<double> rolling_prices; // Sliding window for rolling mean
    double rolling_mean_last = 0.0;    // Latest rolling mean
    double last_price = -1.0;          // Previous price (for anomaly detection)
    double peak_price = -1.0;          // Highest observed price (for drawdown)
    double max_drawdown_pct = 0.0;     // Maximum drawdown percentage
    int anomaly_count = 0;             // Count of detected anomalies

    // Update state with a new tick (price + volume)
    void update(double price, int volume) {
        // Update volume and notional (used for VWAP calculation)
        total_volume += volume;
        total_notional += price * volume;

        // Maintain rolling window of recent prices
        rolling_prices.push_back(price);
        if (static_cast<int>(rolling_prices.size()) > ROLLING_WINDOW) {
            rolling_prices.pop_front();
        }

        // Compute rolling mean
        double rolling_sum = 0.0;
        for (double p : rolling_prices) {
            rolling_sum += p;
        }
        rolling_mean_last = rolling_sum / rolling_prices.size();

        // Detect price anomalies based on percentage jump
        if (last_price > 0.0) {
            double jump_pct = std::abs((price - last_price) / last_price) * 100.0;
            if (jump_pct >= ANOMALY_THRESHOLD_PCT) {
                anomaly_count++;
            }
        }

        // Track peak price for drawdown calculation
        if (price > peak_price) {
            peak_price = price;
        }

        // Compute max drawdown (percentage drop from peak)
        if (peak_price > 0.0) {
            double drawdown_pct = ((peak_price - price) / peak_price) * 100.0;
            if (drawdown_pct > max_drawdown_pct) {
                max_drawdown_pct = drawdown_pct;
            }
        }

        // Store current price for next iteration
        last_price = price;
    }
};

// Parses a CSV line into symbol, price, and volume
// Returns false if parsing fails or data is invalid
bool parse_line(const std::string& line, std::string& symbol, double& price, int& volume) {
    std::stringstream ss(line);
    std::string timestamp;
    std::string price_str;
    std::string volume_str;

    // Extract CSV fields
    if (!std::getline(ss, timestamp, ',')) return false;
    if (!std::getline(ss, symbol, ',')) return false;
    if (!std::getline(ss, price_str, ',')) return false;
    if (!std::getline(ss, volume_str, ',')) return false;

    // Convert string values to numeric types
    try {
        price = std::stod(price_str);
        volume = std::stoi(volume_str);
    } catch (...) {
        return false; // Handle malformed numeric data
    }

    return !symbol.empty();
}

int main(int argc, char* argv[]) {
    // Expect input CSV and output JSON file paths
    if (argc != 3) {
        std::cerr << "Usage: ./pipeline <input_csv> <output_json>\n";
        return 1;
    }

    // Open input file
    std::ifstream input(argv[1]);
    if (!input.is_open()) {
        std::cerr << "Failed to open input file\n";
        return 1;
    }

    // Map to store state per symbol
    std::unordered_map<std::string, SymbolState> states;
    std::string line;

    std::getline(input, line); // Skip header row

    // Process each line (tick data)
    while (std::getline(input, line)) {
        std::string symbol;
        double price = 0.0;
        int volume = 0;

        // Parse and validate line
        if (!parse_line(line, symbol, price, volume)) {
            continue;
        }

        // Update state for the given symbol
        states[symbol].update(price, volume);
    }

    // Open output file
    std::ofstream output(argv[2]);
    if (!output.is_open()) {
        std::cerr << "Failed to open output file\n";
        return 1;
    }

    // Write results in JSON format
    output << "{\n";
    bool first_symbol = true;

    for (const auto& entry : states) {
        const auto& symbol = entry.first;
        const auto& state = entry.second;

        // Compute VWAP (Volume Weighted Average Price)
        double vwap = state.total_volume > 0 ? state.total_notional / state.total_volume : 0.0;

        if (!first_symbol) {
            output << ",\n";
        }
        first_symbol = false;

        output << "  \"" << symbol << "\": {\n";
        output << "    \"total_volume\": " << state.total_volume << ",\n";
        output << std::fixed << std::setprecision(4);
        output << "    \"vwap\": " << vwap << ",\n";
        output << "    \"rolling_mean_last\": " << state.rolling_mean_last << ",\n";
        output << "    \"anomaly_count\": " << state.anomaly_count << ",\n";
        output << "    \"max_drawdown_pct\": " << state.max_drawdown_pct << "\n";
        output << "  }";
    }

    output << "\n}\n";

    // Final summary output
    std::cout << "Processed " << states.size() 
              << " symbols. Output written to " << argv[2] << "\n";

    return 0;
}

File: src/python/pipeline.py
from __future__ import annotations

import csv
import json
import math
import sys
from collections import defaultdict, deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Deque, Dict, Iterable, List


# Number of recent prices used for rolling mean calculation
ROLLING_WINDOW = 3

# Percentage threshold to flag price anomalies (sudden jumps)
ANOMALY_THRESHOLD_PCT = 0.8


@dataclass
class SymbolState:
    """Maintains aggregated state for a single symbol during processing."""
    
    total_volume: int = 0                  # Total traded volume
    total_notional: float = 0.0            # Sum of (price * volume) for VWAP
    rolling_prices: Deque[float] = field(default_factory=lambda: deque(maxlen=ROLLING_WINDOW))
    rolling_mean_last: float = 0.0         # Latest rolling mean value
    last_price: float | None = None        # Previous price (for anomaly detection)
    peak_price: float = -math.inf          # Highest observed price (for drawdown)
    max_drawdown_pct: float = 0.0          # Maximum drawdown percentage
    anomaly_count: int = 0                 # Count of detected anomalies

    def update(self, price: float, volume: int) -> None:
        """Update symbol state with a new tick (price, volume)."""
        
        # Update volume and notional for VWAP calculation
        self.total_volume += volume
        self.total_notional += price * volume

        # Maintain rolling window and compute moving average
        self.rolling_prices.append(price)
        self.rolling_mean_last = round(sum(self.rolling_prices) / len(self.rolling_prices), 4)

        # Detect anomalies based on percentage price change
        if self.last_price is not None and self.last_price != 0:
            price_jump_pct = abs((price - self.last_price) / self.last_price) * 100
            if price_jump_pct >= ANOMALY_THRESHOLD_PCT:
                self.anomaly_count += 1

        # Track peak price for drawdown calculation
        self.peak_price = max(self.peak_price, price)

        # Compute drawdown (drop from peak)
        if self.peak_price > 0:
            drawdown_pct = ((self.peak_price - price) / self.peak_price) * 100
            self.max_drawdown_pct = max(self.max_drawdown_pct, drawdown_pct)

        # Store current price for next iteration
        self.last_price = price

    def to_dict(self) -> Dict[str, float | int]:
        """Convert internal state to a serialisable dictionary."""
        
        vwap = round(self.total_notional / self.total_volume, 4) if self.total_volume else 0.0

        return {
            "total_volume": self.total_volume,
            "vwap": round(vwap, 4),
            "rolling_mean_last": round(self.rolling_mean_last, 4),
            "anomaly_count": self.anomaly_count,
            "max_drawdown_pct": round(self.max_drawdown_pct, 4),
        }


def parse_rows(csv_path: Path) -> Iterable[dict]:
    """Stream and validate rows from a CSV file.

    Skips malformed rows and ensures correct data types.
    """
    with csv_path.open("r", newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        for row in reader:
            # Basic validation of required fields
            if not row.get("timestamp") or not row.get("symbol"):
                continue

            # Convert fields to correct types
            try:
                row["price"] = float(row["price"])
                row["volume"] = int(row["volume"])
            except (TypeError, ValueError):
                continue

            yield row


def process_ticks(csv_path: Path) -> Dict[str, Dict[str, float | int]]:
    """Process tick data and compute per-symbol metrics."""
    
    states: Dict[str, SymbolState] = defaultdict(SymbolState)

    # Update state for each row (tick)
    for row in parse_rows(csv_path):
        states[row["symbol"]].update(price=row["price"], volume=row["volume"])

    # Convert results into serialisable output
    return {symbol: state.to_dict() for symbol, state in sorted(states.items())}


def main(argv: List[str]) -> int:
    """Entry point for CLI execution."""
    
    if len(argv) != 3:
        print("Usage: python3 pipeline.py <input_csv> <output_json>")
        return 1

    input_path = Path(argv[1])
    output_path = Path(argv[2])

    # Run processing pipeline
    results = process_ticks(input_path)

    # Write output as formatted JSON
    output_path.write_text(json.dumps(results, indent=2), encoding="utf-8")

    print(f"Processed {len(results)} symbols. Output written to {output_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
